File: 240625/강력한 폭발/strong-explosion.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <utility>
#include <tuple>
using namespace std;
#define MAX_N 20
#define BOMB_TYPE_NUM 3

int n, ans;
bool bombed[MAX_N][MAX_N];
int bomb_type[MAX_N][MAX_N];
vector<pair<int, int>> bombs;
pair<int, int> bomb_shapes[BOMB_TYPE_NUM+1][5] = {
    {},
    {{-2, 0}, {-1, 0}, {0, 0}, {1, 0}, {2, 0}},
    {{-1, 0}, {0, -1}, {0, 0}, {0, 1}, {1, 0}},
    {{-1, -1}, {-1, 1}, {0, 0}, {1, -1}, {1, 1}}
};

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

void Bomb(int x, int y, int type) {
    for(int i=0; i<5; i++) {
        int dx, dy;
        tie(dx, dy) = bomb_shapes[type][i];

        int nx = x+dx, ny = y+dy;
        if(InRange(nx, ny)) bombed[nx][ny] = true;
    }
}

int Calc() {
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            if(bomb_type[i][j])
                Bomb(i, j, bomb_type[i][j]);
        }
    }

    int cnt = 0;
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            if(bombed[i][j]) {
                cnt++;
                bombed[i][j] = false;
            }
        }
    }

    return cnt;
}

void Simulate(int cnt) {
    if(cnt == (int)bombs.size()) {
        ans = max(ans, Calc());
        return;
    }

    for(int i=1; i<=3; i++) {
        int x, y;
        tie(x, y) = bombs[cnt];

        bomb_type[x][y] = i;
        Simulate(cnt+1);
        bomb_type[x][y] = 0;
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            int bomb_place;
            cin >> bomb_place;
            if(bomb_place) bombs.push_back(make_pair(i, j));
        }
    }

    Simulate(0);

    cout << ans;
    return 0;
}

File: 250925/강력한 폭발/strong-explosion.py
n = int(input())

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

spot = [
    (i, j)
    for i in range(n)
    for j in range(n)
    if grid[i][j] == 1
]

bomb = [
    [],
    [(-2, 0), (-1, 0), (0, 0), (1, 0), (2, 0)],
    [(-1, 0), (0, 1), (0, 0), (1, 0), (0, -1)],
    [(-1, -1), (-1, 1), (0, 0), (1, 1), (1, -1)]
]

ans = 0
visited = [[0]*n for _ in range(n)]

def possible(x, y):
    return 0<=x<n and 0<=y<n

def calc():
    for x, y in spot:
        bn = grid[x][y]

        for dx, dy in bomb[bn]:
            nx, ny = x+dx, y+dy

            if possible(nx, ny):
                visited[nx][ny] = 1
    
    return sum([
        visited[i][j]
        for i in range(n)
        for j in range(n)
    ])



def simulate(cnt):
    global ans, visited

    if cnt == len(spot):
        visited = [[0]*n for _ in range(n)]
        ans = max(ans, calc())
        return

    x, y = spot[cnt]
    for i in range(1, 4):
        grid[x][y] = i
        simulate(cnt+1)




simulate(0)

print(ans)